rpcs3/Emu/Cell/Modules/cellAdec.cpp

```diff
@@ -225,57 +228,735 @@ void fmt_class_string<CellAdecError>::format(std::string& out, u64 arg)
+	cellAdec.notice("_CellAdecCoreOpOpen_lpcm(handle=*0x%x, notifyAuDone=*0x%x, notifyAuDoneArg=*0x%x, notifyPcmOut=*0x%x, notifyPcmOutArg=*0x%x, notifyError=*0x%x, notifyErrorArg=*0x%x, notifySeqDone=*0x%x, notifySeqDoneArg=*0x%x, res=*0x%x)",
 		handle, notifyAuDone, notifyAuDoneArg, notifyPcmOut, notifyPcmOutArg, notifyError, notifyErrorArg, notifySeqDone, notifySeqDoneArg, res);
 
-	fmt::throw_exception("LPCM decoder not implemented, please disable HLE libadec.sprx");
+	return _CellAdecCoreOpOpenExt_lpcm(ppu, handle, notifyAuDone, notifyAuDoneArg, notifyPcmOut, notifyPcmOutArg, notifyError, notifyErrorArg, notifySeqDone, notifySeqDoneArg, res, vm::null);
 }
 
-error_code _CellAdecCoreOpClose_lpcm(vm::ptr<void> handle)
+error_code _CellAdecCoreOpClose_lpcm(ppu_thread& ppu, vm::ptr<LpcmDecContext> handle)
 {
-	cellAdec.todo("_CellAdecCoreOpClose_lpcm(handle=*0x%x)", handle);
+	ppu.state += cpu_flag::wait;
+
+	cellAdec.notice("_CellAdecCoreOpClose_lpcm(handle=*0x%x)", handle);
+
+	if (error_code ret = sys_mutex_lock(ppu, handle->queue_size_mutex, 0); ret != CELL_OK
+		|| (ret = sys_mutex_lock(ppu, handle->queue_mutex, 0)) != CELL_OK)
+	{
+		return ret;
+	}
+
+	if (handle->cmd_queue.empty())
+	{
+		handle->cmd_queue.emplace(LpcmDecCmdType::close);
+
+		if (error_code ret = sys_mutex_unlock(ppu, handle->queue_mutex); ret != CELL_OK)
+		{
+			return ret; // LLE doesn't unlock the queue size mutex
+		}
+
+		if (error_code ret = handle->cmd_available.release(ppu); ret != CELL_OK)
+		{
+			ensure(sys_mutex_unlock(ppu, handle->queue_size_mutex) == CELL_OK); // Error code isn't checked on LLE
+			return ret;
+		}
+	}
+	else
+	{
+		for (auto& cmd : handle->cmd_queue.elements)
+		{
+			cmd.type = LpcmDecCmdType::close;
+		}
+
+		if (error_code ret = sys_mutex_unlock(ppu, handle->queue_mutex); ret != CELL_OK)
+		{
+			return ret; // LLE doesn't unlock the queue size mutex
+		}
+	}
+
+	vm::var<u64> thread_ret;
+
+	if (error_code ret = sys_mutex_unlock(ppu, handle->queue_size_mutex); ret != CELL_OK
+		|| (ret = handle->release_output(ppu)) != CELL_OK
+		|| (ret = sys_ppu_thread_join(ppu, static_cast<u32>(handle->thread_id), +thread_ret)) != CELL_OK
+		|| (ret = sys_cond_destroy(ppu, handle->queue_size_cond)) != CELL_OK
+		|| (ret = sys_cond_destroy(ppu, handle->unk_cond)) != CELL_OK
+		|| (ret = sys_cond_destroy(ppu, handle->output_consumed)) != CELL_OK
+		|| (ret = sys_mutex_destroy(ppu, handle->queue_mutex)) != CELL_OK
+		|| (ret = sys_mutex_destroy(ppu, handle->queue_size_mutex)) != CELL_OK
+		|| (ret = sys_mutex_destroy(ppu, handle->unk_mutex)) != CELL_OK
+		|| (ret = sys_mutex_destroy(ppu, handle->output_mutex)) != CELL_OK
+		|| (ret = handle->cmd_available.finalize(ppu)) != CELL_OK
+		|| (ret = handle->reserved2.finalize(ppu)) != CELL_OK
+		|| (ret = sys_mutex_destroy(ppu, handle->spurs_queue_pop_mutex)) != CELL_OK
+		|| (ret = sys_mutex_destroy(ppu, handle->spurs_queue_push_mutex)) != CELL_OK)
```

Comment:
This if state can be written like instead:

error_code ret{};
ret = ret ? ret : sys_mutex_unlock(ppu...);
ret = ret ? ret : handle->release_output(ppu);

This is more readable IMO